### PR TITLE
yum: documentation change to reflect cases like Fixes #24059

### LINKED ISSUE #24059: Ansible v2.3 centos7 error: TypeError: _execute_module() got an unexpected keyword argument 'wrap_async'
##### ISSUE TYPE
 - Bug Report

##### COMPONENT NAME
Simple yum installations

##### ANSIBLE VERSION
```
ansible 2.3.0.0
  config file = /etc/ansible/ansible.cfg
  configured module search path = Default w/o overrides
  python version = 2.7.5 (default, Nov  6 2016, 00:28:07) [GCC 4.8.5 20150623 (Red Hat 4.8.5-11)]
```

##### CONFIGURATION
See below.

##### OS / ENVIRONMENT
CentOS ansible docker image: https://hub.docker.com/r/geerlingguy/docker-centos7-ansible/

##### SUMMARY
I use docker / [this test script from geerlingguy](https://gist.github.com/geerlingguy/73ef1e5ee45d8694570f334be385e181) to test my roles. Two days ago, everything ran fine. Today I receive the errors you find below.

```
TASK [ldap : install sssd client] **********************************************
task path: /etc/ansible/roles/role_under_test/roles/ldap/tasks/main.yml:3
An exception occurred during task execution. The full traceback is:
Traceback (most recent call last):
  File "/usr/lib/python2.7/site-packages/ansible/executor/task_executor.py", line 126, in run
    display.debug("_execute() done")
  File "/usr/lib/python2.7/site-packages/ansible/executor/task_executor.py", line 502, in _execute
    retries = 1
  File "/usr/lib/python2.7/site-packages/ansible/plugins/action/normal.py", line 45, in run
    results = merge_hash(results, self._execute_module(tmp=tmp, task_vars=task_vars, wrap_async=wrap_async))
TypeError: _execute_module() got an unexpected keyword argument 'wrap_async'

fatal: [localhost]: FAILED! => {
    "failed": true,
    "msg": "Unexpected failure during module execution.",
    "stdout": ""
}
```
Here is the task I run:

```
#install sssd client
- name: install sssd client
  yum: name=sssd state=latest
```

##### STEPS TO REPRODUCE
I created a repo for this bug. Simply clone it and follow the instructions in the README:https://github.com/Kecksdose/ansible-centos7-bug 

```yaml
---
- name: Update yum packages
  yum: name=* state=latest

- name: install sssd client
  yum: name=sssd state=latest
```

##### EXPECTED RESULTS
No errors.

##### ACTUAL RESULTS
```
TASK [test : install sssd client] **********************************************
task path: /etc/ansible/roles/role_under_test/roles/test/tasks/main.yml:5
An exception occurred during task execution. The full traceback is:
Traceback (most recent call last):
  File "/usr/lib/python2.7/site-packages/ansible/executor/task_executor.py", line 126, in run
    display.debug("_execute() done")
  File "/usr/lib/python2.7/site-packages/ansible/executor/task_executor.py", line 502, in _execute
    retries = 1
  File "/usr/lib/python2.7/site-packages/ansible/plugins/action/normal.py", line 45, in run
    results = merge_hash(results, self._execute_module(tmp=tmp, task_vars=task_vars, wrap_async=wrap_async))
TypeError: _execute_module() got an unexpected keyword argument 'wrap_async'

fatal: [localhost]: FAILED! => {
    "failed": true,
    "msg": "Unexpected failure during module execution.",
    "stdout": ""
}
	to retry, use: --limit @/etc/ansible/roles/role_under_test/run_test.retry

PLAY RECAP *********************************************************************
```

### PR DESCRIPTION
**Pull request by ReSearchITEng**
Fixes #24059

##### ISSUE TYPE

 - Feature Pull Request
 - New Module Pull Request
 - Bugfix Pull Request
 - Docs Pull Request

##### ANSIBLE VERSION

```

```

##### ADDITIONAL INFORMATION

```

```

### CODE REVIEW

**Comment by ReSearchITEng** [comment]
Documentation Fixes #24059

**Comment by dagwieers** [comment]
I understand that you think the fix is so obvious, you don't need to fill out the issue-template. However the bug-triage is based on the information you provide. So please fill out the template form the best you can.

File: lib/ansible/modules/packaging/os/yum.py [hunk 1]
```diff
@@ -44,7 +44,7 @@
     aliases: [ 'pkg' ]
   exclude:
     description:
-      - "Package name(s) to exclude when state=present, or latest"
+      - "Package name(s) to exclude when state=present, or latest; Note: updates to ansible binary while a playbook is running can cause issues. It is advisable to exclude ansible when performing os updates from ansible."
```
Comment (dagwieers) [inline lib/ansible/modules/packaging/os/yum.py]: Could you:
- remove the quotes around the complete line
- Move the note to the second item in the list, remove the "Note:"

Alternatively, you can add you suggestion to the `notes:` key at the root of the documentation.

File: lib/ansible/modules/packaging/os/yum.py [hunk 2]
```diff
@@ -193,6 +193,7 @@
   yum:
     name: '*'
     state: latest
+    exclude: 'ansible'
```
Comment (dagwieers) [inline lib/ansible/modules/packaging/os/yum.py]: So here I would say, remove the quotes. They are not needed, and people may think they are needed.

What's more, reviewing the code there is a real shortcoming. You can only exclude one single package this way.
This should be redesigned to be a proper YAML list, so we can simply do:
```yaml
  exclude:
  - ansible
  - yum
```

Are you comfortable to make this change in Python too ?

File: lib/ansible/modules/packaging/os/yum.py [hunk 3]
```diff
@@ -44,7 +44,7 @@
     aliases: [ 'pkg' ]
   exclude:
     description:
-      - "Package name(s) to exclude when state=present, or latest"
```
Comment (gundalow) [inline lib/ansible/modules/packaging/os/yum.py:47]: lib/ansible/modules/packaging/os/yum.py:47:161: E501 line too long (221 > 160 characters)

**Review by gundalow — CHANGES_REQUESTED** [review]

**Review by dagwieers — CHANGES_REQUESTED** [review]

**Comment by abadger** [comment]
I'm not sure that this is the place to document this caveat and whether this it the proper advice to give.  
* Seems like an FAQ because the problem can occur for anything that's updating Ansible.
* The advice should probably be something like, if you're going to update all packages on the system and ansible is installed via packages, it's advisable to update the ansible package separately from everything else.

**Comment by bcoca** [comment]
I agree with @abadger , this is not a yum specific issue ... its mostly common sense, do not replace the screwdriver in the middle of screwing in the plane's fuselage while in flight.

**Comment by dagwieers** [comment]
Despite the note in the description, it would still be useful to have an example showing how to use the exclude parameter (and if this example also clarifies the original issue, I am fine with that).

And for that we definitely want to be able to exclude more than one package, so the feature requested would be nice to fix as well.

File: lib/ansible/modules/packaging/os/yum.py [hunk 4]
```diff
@@ -193,6 +194,7 @@
   yum:
     name: '*'
     state: latest
+    exclude: ansible
```
Comment (dagwieers) [inline lib/ansible/modules/packaging/os/yum.py:197]: Reviewing the code there is a real shortcoming. You can only exclude one single package this way.
This should be redesigned to be a proper YAML list, so we can simply do:
```yaml
  exclude:
  - ansible
  - yum
```

Are you comfortable to make this change in Python too ?

**Review by dagwieers — CHANGES_REQUESTED** [review]

**Comment by abadger** [comment]
Yeah, I agree with dag that documenting exclude is desirable and making exclude a list would be a good feature enhancement.

File: lib/ansible/modules/packaging/os/yum.py (hunk 4 above)
Comment (bcoca) [reply lib/ansible/modules/packaging/os/yum.py:197]: +1 to this

**Comment by maxamillion** [comment]
It appears from the most recent conversation on this pull request that there are changes requested and now as an unfortunate side effect of the elapsed time, there's a merge conflict. Please rebase, resolve the merge conflicts and update based on feedback. Thank you!

**Comment by maxamillion** [comment]
@ReSearchITEng do you still desire to contribute this pull request?

**Comment by maxamillion** [comment]
Closing PR due to inactivity, please feel free to re-open and ping me if you desire to continue working on it.

**Pull request closed without merging**